lion/io/database_parameters.h

```diff
@@ -244,4 +255,27 @@ inline void write_parameters(Xml_document& doc, const std::string& path, const s
     }
 }
 
+bool database_parameters_all_used(Xml_document& doc)
+{
+    bool parameters_all_used = true;
+    Xml_element element = doc.get_root_element();
+    std::vector<Xml_element> children;
+    while(element.has_children())
+    {
+        children = element.get_children();
+        for (Xml_element child : children)
+        {
+            if(!child.has_attribute("__unused__"))
+            {
+                parameters_all_used = false;
+                return parameters_all_used;
+            }
+        }
+
+        if(element.has_sibling()){element = element.get_sibling();}
```

Comment:
I do not  see how this algorithm can work for three levels of children

```xml
<a>
   <b>
      <c> 
          <d/>
      </c>
    </b>
</a>
```

maybe I am wrong but I'd say a recursive algorithm is needed here

Reply:
To make it recursive, for convenience, you can change the prototype to

```C++ 
bool database_parameters_all_used(Xml_element& e)
```

instead to take the `Xml_document`. It can be called by passing the root element `database_parameters_all_used(doc.get_root_element())`

Reply:
We could have this structure:
```cpp
bool database_parameters_all_used(Xml_database& doc)
{
    bool parameters_all_used = true;
    Xml_element element = doc.get_root_element();
    database_parameters_all_used_impl(element, parameters_all_used)
    return parameters_all_used;
}
// Recursive in nature
static void database_parameters_all_used_impl(Xml_element& element, bool& decision);
```

Reply:
I dont get why the implementation function needs to take the decision parameter.

Reply:
I prefer the other prototype I sent you better